tuplex/codegen/src/IteratorContextProxy.cc

```
@@ -0,0 +1,494 @@
+            }
+
+            builder.SetInsertPoint(notExhaustedBB);
+            auto nextVal = getIteratorNextElement(builder, yieldType, iterator, iteratorInfo);
+            llvm::Value *retValNotExhausted = nextVal.val, *retSizeNotExhausted = nextVal.size;
+            builder.CreateBr(endBB);
+
+            builder.SetInsertPoint(defaultArgBB);
+            builder.CreateBr(endBB);
+
+            builder.SetInsertPoint(endBB);
+            lfb.setLastBlock(endBB);
+            if(defaultArg.val) {
+                auto retVal = builder.CreatePHI(_env->pythonToLLVMType(yieldType), 2);
+                auto retSize = builder.CreatePHI(_env->i64Type(), 2);
+                retVal->addIncoming(retValNotExhausted, notExhaustedBB);
+                retSize->addIncoming(retSizeNotExhausted, notExhaustedBB);
+                retVal->addIncoming(defaultArg.val, defaultArgBB);
+                retSize->addIncoming(defaultArg.size, defaultArgBB);
+                return SerializableValue(retVal, retSize);
+            } else {
+                return SerializableValue(retValNotExhausted, retSizeNotExhausted);
+            }
+        }
+
+        llvm::Value *IteratorContextProxy::updateIteratorIndex(llvm::IRBuilder<> &builder,
+                                                               llvm::Value *iterator,
+                                                               IteratorInfo *iteratorInfo) {
+            using namespace llvm;
+
+            llvm::Type *iteratorContextType = iterator->getType()->getPointerElementType();
+            std::string funcName;
+            auto iteratorName = iteratorInfo->iteratorName;
+
+            if(iteratorName == "zip") {
+                return updateZipIndex(builder, iterator, iteratorInfo);
+            }
+
+            if(iteratorName == "enumerate") {
+                return updateEnumerateIndex(builder, iterator, iteratorInfo);
+            }
+
+            assert(iteratorName == "iter");
```

I think this assert should get converted to an if, and whenever an unknown iterator name is used this function should throw an error and use fallback mode.

---

Added error message.